service/controller/v11/resource/ingress/current.go

```diff
@@ -43,5 +45,28 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 
 	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d ingresses in the Kubernetes API", len(ingresses)))
 
+	// In case a cluster deletion happens, we want to delete the guest cluster
+	// ingresses. We still need to use the ingresses for ingress routing in order
+	// to drain nodes on KVM though. So as long as pods are there we delay the
+	// deletion of the ingresses here in order to still be able to route traffic
+	// to the guest cluster API. As soon as the draining was done and the pods got
+	// removed we get an empty list here after the delete event got replayed. Then
+	// we just remove the ingresses as usual.
+	if key.IsDeleted(customObject) {
+		n := key.ClusterNamespace(customObject)
+		list, err := r.k8sClient.CoreV1().Pods(n).List(metav1.ListOptions{})
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		if len(list.Items) != 0 {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "cannot finish deletion of ingresses due to existing pods")
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil, nil
+		}
+	}
```

Comment:
We have to keep the connections to the guest clusters open so we have to sort this in certain resources just like we did in other operators. The amount of modifications is a bit overwhelming to be honest but it makes total sense to make the resources implement this behaviour based on the current state of the system.